shared-src/gradle-plugin/protokt/v1/gradle/ProtobufBuild.kt

```diff
@@ -52,51 +53,90 @@ internal fun configureProtobufPlugin(
         }
 
         generateProtoTasks {
+            val mainExtractProtoAdditions = mutableListOf<TaskInputFiles>()
+            val testExtractProtoAdditions = mutableListOf<TaskInputFiles?>()
+
             for (task in all()) {
                 if (disableJava) {
                     task.builtins {
                         findByName("java")?.run(::remove)
                     }
                 }
 
+                val extensions = resolveExtensions(project, task)
+
+                if (task.isTestTask()) {
+                    testExtractProtoAdditions.add(extensions.test?.let { TaskInputFiles(extensions.taskName, it) })
+                } else {
+                    mainExtractProtoAdditions.add(TaskInputFiles(extensions.taskName, extensions.main))
+                }
+
                 task.plugins {
                     id(target.protocPluginName) {
-                        project.afterEvaluate {
-                            option("$KOTLIN_EXTRA_CLASSPATH=${extraClasspath(project, task)}")
-                            option("$GENERATE_TYPES=${ext.generate.types}")
-                            option("$GENERATE_DESCRIPTORS=${ext.generate.descriptors}")
-                            option("$GENERATE_GRPC_DESCRIPTORS=${ext.generate.grpcDescriptors}")
-                            option("$GENERATE_GRPC_KOTLIN_STUBS=${ext.generate.grpcKotlinStubs}")
-                            option("$FORMAT_OUTPUT=${ext.formatOutput}")
-                            option("$KOTLIN_TARGET=$target")
-                        }
+                        option("$KOTLIN_EXTRA_CLASSPATH=${extraClasspath(project, extensions)}")
+                        option("$GENERATE_TYPES=${ext.generate.types}")
+                        option("$GENERATE_DESCRIPTORS=${ext.generate.descriptors}")
+                        option("$GENERATE_GRPC_DESCRIPTORS=${ext.generate.grpcDescriptors}")
+                        option("$GENERATE_GRPC_KOTLIN_STUBS=${ext.generate.grpcKotlinStubs}")
+                        option("$FORMAT_OUTPUT=${ext.formatOutput}")
+                        option("$KOTLIN_TARGET=$target")
                     }
                 }
             }
+
+            project.handleExtraInputFiles(mainExtractProtoAdditions, testExtractProtoAdditions)
         }
     }
 }
 
-private fun extraClasspath(project: Project, task: GenerateProtoTask): String {
-    var extensions: FileCollection = project.configurations.getByName(EXTENSIONS)
-
-    if (task.isTest) {
```

Comment:
Turns out this isn't reliable - extractIncludeJvmTestProto is considered `main` because it doesn't perfectly match p-g-p's criteria. I haven't looked how deeply that check failing affects our config.